+nosnoc/+solver/MpccSolver.m

```diff
@@ -426,5 +458,415 @@ function print_infeasibility(obj, results)
                 keyboard
             end
         end
+
+        function [polished_w, res_out, stat_type, n_biactive] = calculate_stationarity(obj, exitfast, lifted)
+            import casadi.*
+            stat_type = "?";
+            nlp = obj.nlp;
+            mpcc = obj.mpcc;
+
+            % This method generates a TNLP (tightened NLP) using the results of a homotopy solver.
+            % In order to do so we first identify the active set of the complementarity constraints which involves identifying which
+            % branch of the complementarity each element of G(w) and H(w) are, or whether the solution is biactive at that point.
+            % For a more in depth explanation see the PhD theses of Armin Nurkanovic or Alexandra Schwartz
+            
+            w_orig = nlp.w.mpcc_w().res;
+            lam_x = nlp.w.mpcc_w().mult;
+            lam_g = nlp.g.mpcc_g().mult;
+
+            G = mpcc.G;
+            H = mpcc.H;
+            G_old = full(obj.G_fun(w_orig, nlp.p.mpcc_p().val));
+            H_old = full(obj.H_fun(w_orig, nlp.p.mpcc_p().val));
+            if obj.opts.homotopy_steering_strategy == HomotopySteeringStrategy.DIRECT
+                a_tol = 1*sqrt(obj.opts.comp_tol);
+            else
+                a_tol = 1*sqrt(max(nlp.w.s_elastic().res));
+            end
+            
+            ind_00 = G_old<a_tol & H_old<a_tol;
+            n_biactive = sum(ind_00);
+            if exitfast && n_biactive == 0
+                polished_w = [];
+                res_out = []
+                stat_type = "S";
+                disp("Converged to S-stationary point")
+                return
+            end
+            mindists = max(G_old, H_old);
+            [~, min_idx] = sort(mindists);
+
+            if lifted
```

Comment:
rename lifted to slightly more descriptive name, e.g., complementarity_constraints_lifted